### DISCUSSION TITLE
Can you provide a new example for setting up a SOCKS5 proxy for TCP and UDP traffic for a specified PID?

### DISCUSSION THREAD

**Discussion by 23233** [wiresock/ndisapi#20] [Ideas]
Background Story:
I have several instances of an application running simultaneously, and I need each instance to use a different proxy.

![OO6${4VA$%AW5V ~7ZMSMQX](https://user-images.githubusercontent.com/12870070/232978684-4161cef7-d1a1-47ff-a326-589f96976b42.png)

If possible, please provide a solution.
Thank you very much!

**Comment by wiresock** [comment]
Although this task appears feasible, and I have worked on similar projects in the past, it seems overly complex for the demo. The Socksify demo already includes an abundance of configuration options that need to be set before starting. The demo you're requesting would necessitate setting up an intricate configuration to associate process IDs (PIDs) with SOCKS5 proxy instances, which adds another layer of complexity.
Please note that the examples provided are designed to showcase the capabilities and concepts of the [Windows Packet Filter](https://www.ntkernel.com/windows-packet-filter/), rather than serving as fully-fledged end-user applications. The primary goal is to offer a clear understanding of the application's potential and inspire further development. However, if you are interested in developing such an application as an open-source project, I would be more than happy to provide assistance and support your efforts.

**Reply by 23233** [reply]
Yes, if this were a functional demo, it might be too complex, but before this, I found a demo in netfiltersdk that has similar functionality. I want to move away from netfilter and join you. So, I am seeking your help to provide a natural transition.

Additionally, I would like to emphasize that fast, efficient traffic capture and modification or forwarding is a core part of network-based applications. Thank you for your contributions.

![image](https://user-images.githubusercontent.com/12870070/233054395-1c5b3900-97ab-4d74-84d0-233c0f12932d.png)

**Reply by wiresock** [reply]
There is an example project called **[socksify](https://github.com/wiresock/ndisapi/tree/master/examples/cpp/socksify)**, which can redirect a specified process (selected by its name) through a designated SOCKS5 proxy. This example should provide a solid foundation for getting started. However, it's important to note that the key difference with your requirement is that it redirects traffic based on the process name, rather than the process ID.

**Comment by wiresock** [comment]
I have added an advanced version of socksify demo which supports UDP and multiply SOCKS5 proxies. The repository is located [here](https://github.com/wiresock/socksify)

**Reply by 23233** [reply]
Oh my goodness, while I was busy preparing and didn't even notice, you have already finished writing. Thank you so much. Once I study and take a look at it, I will open-source my application as a way to give back to you. Thank you!

**Reply by mingtianquan** [reply]
https://github.com/wiresock/proxifyre  这个示例可以pid 设置 TCP 和 UDP 流量的 SOCKS5 代理？

**Reply by wiresock** [reply]
> https://github.com/wiresock/proxifyre  这个示例可以pid 设置 TCP 和 UDP 流量的 SOCKS5 代理？

ProxiFyre 通过名称筛选进程，但可以修改为按进程 ID 进行筛选。

**Reply by mingtianquan** [reply]
大佬怎么咋个修改为pid来转发流量

**Reply by wiresock** [reply]
> 大佬怎么咋个修改为pid来转发流量

这不是几句话能回答的问题。不过源代码是公开的，你可以在那里找到所有的细节。

**Reply by mingtianquan** [reply]
另外能支持线程调用吗比如同时设置不同的进程，使用不同的socks5

**Reply by wiresock** [reply]
> 另外能支持线程调用吗比如同时设置不同的进程，使用不同的socks5

是的，这是可能的。代码可以将不同的进程（按名称选择）映射到不同的 SOCKS5 代理。
